code/game/machinery/droneDispenser.dm

```diff
@@ -126,114 +155,158 @@
+
+	// We are currently in the middle of something
+	if(timer > world.time)
 		return
-	if(droneMadeRecently)
+
+	switch(mode)
+		if(DRONE_READY)
+			// If we have X drone shells already on our turf
+			if(maximum_idle && (count_shells() >= maximum_idle))
+				return // then do nothing; check again next tick
+			if(begin_create_message)
+				visible_message("<span class='notice'>\
+					[src] [begin_create_message]</span>")
+			if(work_sound)
+				playsound(src, work_sound, 50, 1)
+			mode = DRONE_PRODUCTION
+			timer = world.time + production_time
+			update_icon()
+
+		if(DRONE_PRODUCTION)
+			materials.use_amount(using_materials)
+			if(power_used)
+				use_power(power_used)
+
+			new dispense_type(loc)
+
+			if(create_sound)
+				playsound(src, create_sound, 50, 1)
+			if(end_create_message)
+				visible_message("<span class='notice'>[src] \
+					[end_create_message]</span>")
+
+			mode = DRONE_RECHARGING
+			timer = world.time + cooldownTime
+			update_icon()
+
+		if(DRONE_RECHARGING)
+			if(recharge_sound)
+				playsound(src, recharge_sound, 50, 1)
+			if(recharge_message)
+				visible_message("<span class='notice'>\
+					[src] [recharge_message]</span>")
+
+			mode = DRONE_READY
+			update_icon()
+
+/obj/machinery/droneDispenser/proc/count_shells()
+	. = 0
+	for(var/a in loc)
+		if(istype(a, dispense_type))
+			.++
+
+/obj/machinery/droneDispenser/update_icon()
+	if(stat & (BROKEN|NOPOWER))
+		icon_state = icon_off
+	else if(mode == DRONE_RECHARGING)
 		icon_state = icon_recharging
-		return
-	droneMadeRecently = 1
-	if(begin_create_message)
-		visible_message("<span class='notice'>[src] [begin_create_message]</span>")
-	if(work_sound)
-		playsound(src, work_sound, 50, 1)
-	icon_state = icon_creating
-	sleep(30)
-	icon_state = icon_on
-	materials.use_amount(using_materials)
-	if(power_used)
-		use_power(power_used)
-	new dispense_type(loc)
-	if(create_sound)
-		playsound(src, create_sound, 50, 1)
-	if(end_create_message)
-		visible_message("<span class='notice'>[src] [end_create_message]</span>")
-	icon_state = icon_recharging
-	sleep(cooldownTime)
-	if(!(stat & BROKEN))
-		icon_state = icon_on
+	else if(mode == DRONE_PRODUCTION)
+		icon_state = icon_creating
 	else
-		icon_state = icon_off
-	droneMadeRecently = 0
-	if(recharge_sound)
-		playsound(src, recharge_sound, 50, 1)
-	if(recharge_message)
-		visible_message("<span class='notice'>[src] [recharge_message]</span>")
+		icon_state = icon_on
 
 /obj/machinery/droneDispenser/attackby(obj/item/O, mob/living/user)
 	if(istype(O, /obj/item/stack))
 		if(!O.materials[MAT_METAL] && !O.materials[MAT_GLASS])
 			return ..()
 		if(!metal_cost && !glass_cost)
-			user << "<span class='warning'>There isn't a place to insert [O]!</span>"
-			return
-		var/stack = 1
-		var/obj/item/stack/sheets
-		sheets = O
-		stack = (input(user, "How many sheets do you want to add?.", "Drone Dispenser", "[stack]") as num)
-		if(stack <= 0)
-			return
-		if(!user.canUseTopic(src))
```

Comment:
Given that it doesn't input the user for sheet amount and just grabs all of the sheets that its given, I don't think its needed.